scripts/generators/state_tracker_helper_generator.py

```diff
@@ -0,0 +1,215 @@
+        dictionary[name].add(value)
+
+    def generateHeader(self):
+        # Map feature names to structs that have them
+        featureMap = dict()
+        for struct, info in self.vk.structs.items():
+            if featuresStructPattern.match(struct) is None:
+                continue
+
+            suffix = ''
+            if struct in self.identical_but_different_features:
+                suffix = self.identical_but_different_features[struct]
+
+            for feature in info.members:
+                if feature.type == 'VkBool32':
+                    self.addToDict(featureMap, feature.name + suffix, struct)
+
+        # Generate a comment for every feature regarding where it may be coming from, then sort the
+        # features by that comment.  That ensures features of the same struct end up together.
+        featuresAndOrigins = sorted([(', '.join(sorted(structs)), feature)
+                                     for feature, structs in featureMap.items()])
+
+        out = []
+        out.append('''
+            #pragma once
+
+            // Union of all features defined in VkPhysicalDevice*Features* structs
+            struct DeviceFeatures {
+            ''')
+        for origins, feature in featuresAndOrigins:
+            out.append(f'// {origins}\n')
+            out.append(f'bool {feature};\n')
+
+        out.append('''};
+
+            void GetEnabledDeviceFeatures(const VkDeviceCreateInfo *pCreateInfo, DeviceFeatures *features, const APIVersion &api_version);
+            ''')
+        self.write("".join(out))
+
+    def generateSource(self):
+        out = []
+        # Handle pCreateInfo->pEnabledFeatures first, it's the only one not necessarily found in the
+        # pNext chain.
+        out.append('''
+            #include "state_tracker/state_tracker.h"
+
+            void GetEnabledDeviceFeatures(const VkDeviceCreateInfo *pCreateInfo, DeviceFeatures *features, const APIVersion &api_version) {
+                // Initialize all to false
+                *features = {};
+
+                // handle VkPhysicalDeviceFeatures specially as it is not part of the pNext chain,
+                // and when it is (through VkPhysicalDeviceFeatures2), it requires an extra indirection.
+                const VkPhysicalDeviceFeatures *core_features = pCreateInfo->pEnabledFeatures;
+                if (core_features == nullptr) {
+                    const VkPhysicalDeviceFeatures2 *features2 = vku::FindStructInPNextChain<VkPhysicalDeviceFeatures2>(pCreateInfo->pNext);
+                    if (features2 != nullptr) {
+                        core_features = &features2->features;
+                    }
+                }
+                if (core_features != nullptr) {
+            ''')
+        for member in self.vk.structs['VkPhysicalDeviceFeatures'].members:
+            if member.type == 'VkBool32':
+                out.append(f'features->{member.name} = core_features->{member.name} == VK_TRUE;\n')
+        out.append('''}
```

Comment:
nit:

```suggestion
        out.append('}\n')
```

should do the same thing and easier to read I feel

Reply:
Done